typescript/web/src/connectors/apollo-client/cache-config.ts

```diff
@@ -0,0 +1,104 @@
+import {
+  InMemoryCacheConfig,
+  FieldFunctionOptions,
+  FieldReadFunction,
+  FieldMergeFunction,
+} from "@apollo/client";
+import type { QueryImagesArgs, Image } from "@labelflow/graphql-types";
+
+export type ImagesQueryCache = {
+  images: Image[];
+  totalCount: number | undefined;
+};
+
+const readQueryImages: FieldReadFunction<ImagesQueryCache, Image[]> = (
+  existing,
+  options: FieldFunctionOptions<QueryImagesArgs, Record<string, any>>
+) => {
+  // A read function should always return undefined if existing is
+  // undefined. Returning undefined signals that the field is
+  // missing from the cache, which instructs Apollo Client to
+  // fetch its value from your GraphQL server.
+  if (existing === undefined) {
+    return undefined;
+  }
+  const offset = options.args?.skip || 0;
+  const limit = options.args?.first || undefined;
+  const result =
+    offset > 0 || limit !== undefined
+      ? existing.images.slice(
+          offset,
+          limit !== undefined ? offset + limit : undefined
+        )
+      : existing.images;
+  if (
+    result.length === 0 ||
+    (result as (Image | undefined)[]).includes(undefined) ||
+    // Invalidate cache if returned result is smaller than the limit
+    // but does not match the expected total count of images in the
+    // dataset
+    (((limit && result.length < limit) || !limit) &&
+      (!existing.totalCount || existing.totalCount !== offset + result.length))
+  ) {
+    return undefined;
+  }
+  return result;
+};
+
+const mergeQueryImages: FieldMergeFunction<ImagesQueryCache, Image[]> = (
+  existing,
+  incoming,
+  options: FieldFunctionOptions<QueryImagesArgs, Record<string, any>>
+) => {
+  // Slicing is necessary because the existing data is
+  // immutable, and frozen in development.
+  const merged = existing
+    ? {
+        images: existing.images.slice(0),
+        totalCount: existing.totalCount,
+      }
+    : { images: [], totalCount: undefined };
+  const offset = options.args?.skip || 0;
+  const limit = options.args?.first || undefined;
+  for (let i = 0; i < incoming.length; i += 1) {
+    merged.images[offset + i] = incoming[i];
+  }
```

Comment:
_No really a change-request but rather a recommendation._
For such cases, I'd use [immutability-helper](https://www.npmjs.com/package/immutability-helper) which can, in a single operation, take care of making a deep copy (_but without being subject to object properties changes_) and setting only the updated values in the array. It's been made exactly to cover such cases (_it was created by the React team to help updating deep states_)

Reply:
I don't think immutability-helper would be of much help in that situation, in which I need to offset the indexes beforehand anyway.